src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/IncludeRewriter.cs

```diff
@@ -55,12 +54,66 @@ public override Expression VisitSqlRoot(SqlRootExpression expression, object con
                 }
             }
 
-            if (containsInclude)
+            reorderedExpressions.Add(IncludeUnionAllExpression);
+            return new SqlRootExpression(reorderedExpressions, expression.DenormalizedExpressions);
+        }
+
+        private class TableExpressionComparer : IComparer<TableExpression>
+        {
+            public int Compare(TableExpression x, TableExpression y)
             {
-                reorderedExpressions.Add(IncludeUnionAllExpression);
-            }
+                if (x.SearchParameterQueryGenerator is IncludeQueryGenerator)
+                {
+                    if (!(y.SearchParameterQueryGenerator is IncludeQueryGenerator))
+                    {
+                        return -1;
+                    }
 
-            return new SqlRootExpression(reorderedExpressions, expression.DenormalizedExpressions);
+                    // Both expressions are Include expressions
+                    var xInclude = (IncludeExpression)x.NormalizedPredicate;
+                    var yInclude = (IncludeExpression)y.NormalizedPredicate;
+
+                    if (!xInclude.Iterate && yInclude.Iterate)
+                    {
+                        return 1;
+                    }
+
+                    if (xInclude.Iterate && !yInclude.Iterate)
+                    {
+                        return -1;
+                    }
+
+                    // Both expressions are Include:iterate expressions, order so that _include:iterate source type will appear after relevant include target
+                    var xTargetTypes = !string.IsNullOrEmpty(xInclude.TargetResourceType) ? new List<string>() { xInclude.TargetResourceType } : xInclude.ReferenceSearchParameter.TargetResourceTypes;
+                    var yTargetTypes = !string.IsNullOrEmpty(yInclude.TargetResourceType) ? new List<string>() { yInclude.TargetResourceType } : yInclude.ReferenceSearchParameter.TargetResourceTypes;
```

Comment:
Please avoid allocating inside a comparer method

Reply:
Is there a known performance issue specific to this case?
I'll rewrite it without allocating a new list, but note that it will add several if cases.